src/Disks/ObjectStorages/Encrypted/registerDiskEncryptedOS.cpp

```diff
@@ -0,0 +1,135 @@
+        ret->current_key = current_key;
+        ret->current_key_fingerprint = FileEncryption::calculateKeyFingerprint(current_key);
+        ret->current_algorithm = FileEncryption::getCurrentAlgorithmFromConfig(config, config_prefix);
+
+        FileEncryption::checkKeySize(ret->current_key.size(), ret->current_algorithm);
+
+        auto wrapped_disk_name = config.getString(config_prefix + ".disk", "");
+        if (wrapped_disk_name.empty())
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Disk EncryptedOS requires `disk` field in config");
+
+        auto disk_it = disk_map.find(wrapped_disk_name);
+        if (disk_it == disk_map.end())
+        {
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Cannot wrap disk `{}` with encryption layer `{}`: there is no such disk (it should be initialized before encryption disk)",
+                wrapped_disk_name,
+                disk_name);
+        }
+        ret->wrapped_disk = disk_it->second;
+        if (!dynamic_cast<const DiskObjectStorage *>(ret->wrapped_disk.get()))
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Cannot wrap disk `{}` with encryption layer `{}`: encrypted_os disk is allowed only on top of object storage",
+                wrapped_disk_name,
+                disk_name);
+
+        auto header_cache_path = config.getString(config_prefix + ".header_cache_path", "");
+        if (!header_cache_path.empty())
+        {
+            FileCacheSettings file_cache_settings;
+            file_cache_settings.base_path = header_cache_path;
+
+            if (!config.has(config_prefix + ".header_cache_max_size"))
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Expected header cache size (`header_cache_max_size`) in configuration");
+            if (config.has(config_prefix + ".header_cache_max_size"))
+                file_cache_settings.max_size = parseWithSizeSuffix<uint64_t>(config.getString(config_prefix + ".header_cache_max_size"));
+            if (file_cache_settings.max_size == 0)
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Expected non-zero size for header cache configuration");
+            file_cache_settings.cache_on_write_operations = config.getUInt64(config_prefix + ".cache_header_on_write", false);
+            ret->cache_header_on_write = file_cache_settings.cache_on_write_operations;
+            file_cache_settings.max_file_segment_size = FileEncryption::Header::kSize;
+            ret->header_cache = FileCacheFactory::instance().getOrCreate(disk_name + "_header_cache", file_cache_settings);
+            ret->header_cache->initialize();
+        }
+        return ret;
+    }
+    catch (Exception & e)
+    {
+        e.addMessage("Disk " + disk_name);
+        throw;
+    }
+}
+
+void registerDiskEncryptedOS(DiskFactory & factory, bool /* global_skip_access_check */)
```

Comment:
I just had a thought thought though - instead of adding a new disk type, we could just check in `registerDiskEncrypted` that we are creating encrypted disk on top of `DiskObjectStorage` and in this case we do `disk_object_storage->wrapWithEncryption()` and get `DiskObjectStorage(EncryptedObjectStorage(ObjectStorage))` instead of creating `DiskEncrypted(DiskObjectStorage)` . The only point here - current implementation of `encryptedObjectStorage` forces unencrypted cache, need to make it support encrypted.

Reply:
Ok, let's do this way for now.

Reply:
> I'm not sure how to deal with the fact that the 2 types of disk encryption don't have exactly the same set of configuration options. Do you think it would be okay to have a single "disk type" with a different set of options (e.g. header_cache_path/header_cache_max_size vs path)

hm, looks like I missed this part. I assume the configuration options should be the same. Why are they different? Will check the code.